File: frameworks/AgentsMesh/clients/web/src/lib/api/facade/apikey.ts
// Connect-RPC adapter for proto.apikey.v1.ApiKeyService.
//
// Wire layer is proto-SSOT: returns and consumes `@proto/apikey/v1`
// types directly. No adapter DTO layer — hooks/components consume proto
// types (camelCase, bigint id) as-is.
//
// PR #345 lineage: `createApiKey` returns `{ apiKey, rawKey }` (multi-field
// per conventions §9 exception). The wire carries both tag 1 and tag 2;
// the secret cannot be silently dropped on the wasm hop.
import {
  ApiKeySchema,
  CreateApiKeyRequestSchema,
  CreateApiKeyResponseSchema,
  DeleteApiKeyRequestSchema,
  DeleteApiKeyResponseSchema,
  GetApiKeyRequestSchema,
  ListApiKeysRequestSchema,
  ListApiKeysResponseSchema,
  RevokeApiKeyRequestSchema,
  RevokeApiKeyResponseSchema,
  UpdateApiKeyRequestSchema,
  type ApiKey,
} from "@proto/apikey/v1/api_key_pb";
import { create, toBinary, fromBinary } from "@bufbuild/protobuf";
import { getApiKeyService } from "@/lib/wasm-core";

export type { ApiKey } from "@proto/apikey/v1/api_key_pb";

export async function listApiKeys(
  orgSlug: string,
  opts: { offset?: number; limit?: number } = {},
): Promise<{ items: ApiKey[]; total: number; limit: number; offset: number }> {
  const req = create(ListApiKeysRequestSchema, {
    orgSlug,
    offset: opts.offset,
    limit: opts.limit,
  });
  const bytes = toBinary(ListApiKeysRequestSchema, req);
  const respBytes = await getApiKeyService().listApiKeysConnect(bytes);
  const resp = fromBinary(ListApiKeysResponseSchema, new Uint8Array(respBytes));
  return {
    items: resp.items,
    total: Number(resp.total),
    limit: resp.limit,
    offset: resp.offset,
  };
}

export async function getApiKey(orgSlug: string, id: bigint): Promise<ApiKey> {
  const req = create(GetApiKeyRequestSchema, { orgSlug, id });
  const bytes = toBinary(GetApiKeyRequestSchema, req);
  const respBytes = await getApiKeyService().getApiKeyConnect(bytes);
  return fromBinary(ApiKeySchema, new Uint8Array(respBytes));
}

export interface CreateApiKeyResult {
  apiKey: ApiKey;
  rawKey: string;
}

export async function createApiKey(
  orgSlug: string,
  input: {
    name?: string;
    description?: string;
    scopes?: string[];
    expiresIn?: bigint;
  },
): Promise<CreateApiKeyResult> {
  const req = create(CreateApiKeyRequestSchema, {
    orgSlug,
    name: input.name,
    description: input.description,
    scopes: input.scopes ?? [],
    expiresIn: input.expiresIn,
  });
  const bytes = toBinary(CreateApiKeyRequestSchema, req);
  const respBytes = await getApiKeyService().createApiKeyConnect(bytes);
  const resp = fromBinary(CreateApiKeyResponseSchema, new Uint8Array(respBytes));
  if (!resp.apiKey) {
    throw new Error("createApiKey response missing api_key (proto tag 1)");
  }
  return { apiKey: resp.apiKey, rawKey: resp.rawKey };
}

export async function updateApiKey(
  orgSlug: string,
  id: bigint,
  input: {
    name?: string;
    description?: string;
    scopes?: string[];
    isEnabled?: boolean;
  },
): Promise<ApiKey> {
  const req = create(UpdateApiKeyRequestSchema, {
    orgSlug,
    id,
    name: input.name,
    description: input.description,
    scopes: input.scopes ?? [],
    isEnabled: input.isEnabled,
  });
  const bytes = toBinary(UpdateApiKeyRequestSchema, req);
  const respBytes = await getApiKeyService().updateApiKeyConnect(bytes);
  return fromBinary(ApiKeySchema, new Uint8Array(respBytes));
}

export async function revokeApiKey(orgSlug: string, id: bigint): Promise<void> {
  const req = create(RevokeApiKeyRequestSchema, { orgSlug, id });
  const bytes = toBinary(RevokeApiKeyRequestSchema, req);
  const respBytes = await getApiKeyService().revokeApiKeyConnect(bytes);
  fromBinary(RevokeApiKeyResponseSchema, new Uint8Array(respBytes));
}

export async function deleteApiKey(orgSlug: string, id: bigint): Promise<void> {
  const req = create(DeleteApiKeyRequestSchema, { orgSlug, id });
  const bytes = toBinary(DeleteApiKeyRequestSchema, req);
  const respBytes = await getApiKeyService().deleteApiKeyConnect(bytes);
  fromBinary(DeleteApiKeyResponseSchema, new Uint8Array(respBytes));
}

File: frameworks/AgentsMesh/clients/core/crates/api-client/src/modules/apikey.rs
use crate::ApiClient;
use crate::connect_call::connect_call;
use crate::error::ApiError;
use agentsmesh_types::proto_apikey_v1 as apikey_proto;

// Connect-RPC (binary wire). See proto-naming-conventions.md §2.5.
// These methods call the Connect handlers in
// backend/internal/api/connect/apikey/. Procedure paths derive from
// `proto.apikey.v1.ApiKeyService.<Method>` (conventions §12).
//
// PR #345 lineage: `create_api_key_connect` returns
// `CreateApiKeyResponse {api_key, raw_key}` — the multi-field exception
// per conventions §9. The raw key is now structurally pinned to tag 2;
// the wrapper-stripping bug class cannot recur.

impl ApiClient {
    pub async fn list_api_keys_connect(
        &self,
        req: &apikey_proto::ListApiKeysRequest,
    ) -> Result<apikey_proto::ListApiKeysResponse, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/ListApiKeys",
            req,
        )
        .await
    }

    pub async fn get_api_key_connect(
        &self,
        req: &apikey_proto::GetApiKeyRequest,
    ) -> Result<apikey_proto::ApiKey, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/GetApiKey",
            req,
        )
        .await
    }

    pub async fn create_api_key_connect(
        &self,
        req: &apikey_proto::CreateApiKeyRequest,
    ) -> Result<apikey_proto::CreateApiKeyResponse, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/CreateApiKey",
            req,
        )
        .await
    }

    pub async fn update_api_key_connect(
        &self,
        req: &apikey_proto::UpdateApiKeyRequest,
    ) -> Result<apikey_proto::ApiKey, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/UpdateApiKey",
            req,
        )
        .await
    }

    pub async fn revoke_api_key_connect(
        &self,
        req: &apikey_proto::RevokeApiKeyRequest,
    ) -> Result<apikey_proto::RevokeApiKeyResponse, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/RevokeApiKey",
            req,
        )
        .await
    }

    pub async fn delete_api_key_connect(
        &self,
        req: &apikey_proto::DeleteApiKeyRequest,
    ) -> Result<apikey_proto::DeleteApiKeyResponse, ApiError> {
        connect_call(
            self,
            "/proto.apikey.v1.ApiKeyService/DeleteApiKey",
            req,
        )
        .await
    }
}
